cypress/support/PersistentLogs.js

```diff
@@ -16,15 +21,98 @@ console.log = function (message) {
+// we want to ensure the logging we do here is the final logging for the test 
+// case.
+//
+// Also important to notice is that this results in a call to cy.writeFile
+// which will execute in a different time sequence than all of the other 
+// JavaScript in this code block. Thus in this code block we cannot run any
+// true `beforeEach` code that is related to the current test case that is
+// just about to start.
+// FINAL LOGGING SUPPORT FOR PREVIOUS TEST CASE
+beforeEach(function() {
+  LogTestState()
+  FlushLogEntries()
+})
+
+// TRUE `beforeEach` CODE FOR THE TEST CASE THAT IS JUST NOW STARTING
+// This MUST come after the prior `beforeEach` code block and cannot
+// be combined with that other code block due to the `cy.writeFile`
+// call that it contains.
+beforeEach(function() {
+  console.log('')
+  console.log(`******************** ${GetFullTestSuiteTitle(this.currentTest)} ************************************************************`)
+
+  test = this.currentTest
+  let title = GetSuiteRootTitle(this.currentTest)
+  if (title !== suiteTitle) {
+    suiteTitle = title
+    logFileName = `./results/cypress/${suiteTitle}.${Cypress.moment().format("YY.MM.DD.HH.mm.ss..SSS")}.log`
+  }
+})
 
 // After each test case ends...
-afterEach(() => {
-  let logFileName = `./results/cypress/${testCaseRegistry[indexRegistry++]}.${Cypress.moment().format("YY.MM.DD.HH.mm.ss..SSS")}.log`
-  cy.writeFile(logFileName, logEntries).then(() => logEntries = '')
+afterEach(() => { WriteLogEntries() })
+
+after(() => { 
+  LogTestState()
+  FlushLogEntries() 
 })
+
+// -----------------------------------------------------------------------------
+// Support functions.
+// -----------------------------------------------------------------------------
+
+function LogTestState()
+{
+  if (!test) { return false }
+
+  let message = `: ${test.state.toUpperCase()} - Test Case: '${GetFullTestSuiteTitle(test)}' - ${test.state.toUpperCase()} :`
+  console.log('.'.repeat(message.length))
+  console.log(message)
+  console.log("'".repeat(message.length))
+}
+
+// NOTE: This requires the latest version (3.1.5) of Cypress:
+//        change directory to the root project folder
+//        npm install cypress --save-dev
+function WriteLogEntries() { 
+  if (logFileName) {
+    cy.writeFile(logFileName, logEntries, {flag: 'a'})
+      .then(() => { logEntries = '' })
+  }
+} 
+
+function FlushLogEntries() {   
+  if (logFileName && logEntries.length > 0) {
+    // Because of 'afterEach' functions that were called after we wrote our previous
+    // log entries, there are additional log entries to persist to that same log file.
+    WriteLogEntries()
+  }
+}
+
+function GetSuiteRootTitle(test) {
+  if (!test.parent) throw new 'Did not find test.parent'
+  
+  if (test.parent.title === '') { return test.title }
+  return GetSuiteRootTitle(test.parent)
+}
+
+function GetFullTestSuiteTitle(test) {
+  if (!test.parent) throw new 'Did not find test.parent'
```

Comment:
Seems to repeat the same code as above. Is there a way to combine these?

Reply:
In one method I need just the root parent title, this is for the log file name and knowing when I need to start a new log file.

In the other case I need the full chain of titles with a dash separating them for visual readability since they appear in the logs to separate test stages in the log.

I sent you an example of a log file on teams. It will help you understand why I need the code the way it is. I don't think combining the functionality into one function would be nice, it can be done, but I need two different returns from each for different purposes. The function names state the purpose and make the code more readable.